core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -352,54 +358,81 @@ public static String sha512(ByteString string)  {
     return DigestUtils.sha512Hex(string.getBytes());
   }
 
-  /** SQL {@code REGEXP_CONTAINS(value, regexp)} function.
-   * Throws a runtime exception for invalid regular expressions.*/
-  public static boolean regexpContains(String value, String regex) {
-    try {
-      // Uses java.util.regex as a standard for regex processing
-      // in Calcite instead of RE2 used by BigQuery/GoogleSQL
-      Pattern regexp = Pattern.compile(regex);
-      return regexp.matcher(value).find();
-    } catch (PatternSyntaxException ex) {
-      throw RESOURCE.invalidInputForRegexpContains(ex.getMessage().replace("\r\n", " ")
-          .replace("\n", " ").replace("\r", " ")).ex();
+  /** State for {@code REGEXP_CONTAINS}, {@code REGEXP_REPLACE}, {@code RLIKE}.
+   *
+   * <p>Marked deterministic so that the code generator instantiates one once
+   * per query, not once per row. */
+  @Deterministic
+  public static class RegexFunction {
+    private final LoadingCache<Ord<String>, Pattern> cache =
+        CacheBuilder.newBuilder().build(
+            new CacheLoader<Ord<String>, Pattern>() {
+              @SuppressWarnings("MagicConstant")
+              @Override public Pattern load(Ord<String> key) {
+                final String regex = key.e;
+                final int flags = key.i;
+                return Pattern.compile(regex, flags);
+              }
+            });
+
+    /** SQL {@code REGEXP_CONTAINS(value, regexp)} function.
+     * Throws a runtime exception for invalid regular expressions.*/
+    public boolean regexpContains(String value, String regex) {
+      final Pattern pattern;
+      try {
+        // Uses java.util.regex as a standard for regex processing
+        // in Calcite instead of RE2 used by BigQuery/GoogleSQL
+        pattern = cache.getUnchecked(Ord.of(0, regex));
+      } catch (UncheckedExecutionException e) {
```

Comment:
It seems that only `regexpContains` catch this UncheckedExecutionException. other regex or like functions without this behavior. they throw pre-defined error msg in the parameters' checking.

Reply:
This is the only one where I needed to catch the exception, because this is the only one that had negative tests (i.e. do we get a good error message if the regex is invalid).

If the functions do not have sufficient negative tests, I do not intend to fix that in this commit.